onnx/defs/shape_inference.h

```diff
@@ -313,11 +313,15 @@ inline void propagateShapeFromInputToOutput(
 }
 
 inline void propagateShapeAndTypeFromFirstInput(InferenceContext& ctx) {
-  propagateElemTypeFromInputToOutput(ctx, 0, 0);
+  for (int i = 0; i < static_cast<int>(ctx.getNumOutputs()); ++i) {
```

Comment:
This changes the behavior for all existing uses of this function. I think it would be better to explicitly indicate whether we want to propagage just 0 => 0 or 0 => all. What if we add a Boolean parameter to the function (with a default value) to choose between the two? Or, alternatively, take an integer parameter N to indicate propagation from first input to the first N outputs (with a negative value to indicate all)?

Reply:
In practice this only affects `BatchNormalization` which has multiple outputs and uses this. All other uses have only 1 output.

What about: Have this function take an optional param **to**  which output it should be propagated. Additionally add another function `propagateShapeAndTypeFromFirstInputToAllOutputs` which will be used by `Dropout`. 

Reasoning: Using parameters the meaning gets obscured.   `propagateShapeAndTypeFromFirstInput(ctx, 2)` can probably still be understood as `to output 2` and this is more useful than an "all-or-nothing" parameter. The "all" function can then be used directly assuming this might be quite common (more common than to only a subset which can still be implemented by means of the base function)

Reply:
That sounds fine.

Reply:
Ok done. Had to do some minor changes to that, so best see the commits and messages please